### PR TITLE
 [opengl] Detect driver existance in with_opengl()

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #664

As you mentioned in https://github.com/taichi-dev/taichi/pull/843#issuecomment-617544821.

We're on the edge of v0.6!
In this PR, I do `initialize_opengl` in `with_opengl` so that we can detect if driver available, and compute shaders are supported.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/backends/opengl/opengl_api.cpp [hunk 1]
```diff
@@ -243,11 +243,19 @@ struct GLSSBO {
...
   static bool gl_inited = false;
-  if (gl_inited)
-    return;
-  gl_inited = true;
+  static bool gl_supp = false;
+
+
+  if (gl_inited) {
+    if (!gl_supp) {
+      TI_ASSERT_INFO(noerr, "OpenGL not supported");
```
Comment (yuanming-hu) [inline taichi/backends/opengl/opengl_api.cpp:253]: I don't understand this. If `noerr == false`, which means errors are allowed, then this assertion will fail?

File: taichi/backends/opengl/opengl_api.cpp [hunk 2]
```diff
@@ -277,8 +289,16 @@ void initialize_opengl() {
...
-  if (!opengl_has_GL_ARB_compute_shader)
+  if (!opengl_has_GL_ARB_compute_shader) {
+    if (noerr) {
+      TI_INFO("Your OpenGL does not support GL_ARB_compute_shader extension");
+      return false;
+    }
     TI_ERROR("Your OpenGL does not support GL_ARB_compute_shader extension");
+  }
+
+  gl_supp = true;
```
Comment (yuanming-hu) [inline taichi/backends/opengl/opengl_api.cpp:300]: Should here be `gl_init`?

File: taichi/backends/opengl/opengl_api.cpp (hunk 1 above)
Comment (archibate) [reply taichi/backends/opengl/opengl_api.cpp:253]: Oh seems I've made a bad name, actually I mean `noerr -> no_raising_error`, and thus `noerr == true` means errors are allowed.
Sorry about the confusing name :) Maybe we want `noerr -> error_tolerance`?

File: taichi/backends/opengl/opengl_api.cpp (hunk 2 above)
Comment (archibate) [reply taichi/backends/opengl/opengl_api.cpp:300]: Let's use `std::optional` to make the logic here better.

**Comment by archibate** [comment]
Why is Travis still running from yesterday to today...how slow it is...

**Comment by yuanming-hu** [comment]
> Why is Travis still running from yesterday to today...how slow it is...

I triggered a re-build just now - let's see if it passes now :-)

**Pull request merged**